backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt

```diff
@@ -65,13 +65,13 @@ val IrClass.objectIsShared get() =
 
 internal val IrField.storageClass: FieldStorage get() {
     val descriptor = propertyDescriptor
+    val isVal = this.isFinal
     return when {
         descriptor.annotations.hasAnnotation(threadLocalAnnotationFqName) -> FieldStorage.THREAD_LOCAL
-        descriptor.annotations.hasAnnotation(sharedAnnotationFqName) -> FieldStorage.SHARED
+        isVal && descriptor.annotations.hasAnnotation(sharedAnnotationFqName) -> FieldStorage.SHARED
```

Comment:
So `isVal` check can be extracted to the separate branch above, and I believe that the entire strategy would sound closer to what we'd like to say.